File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        i, j = 0, 0
        
        for k in range(len(prices)):
            if prices[k] > prices[j]:
                j = k
                res = max(res, prices[k] - prices[i])
            elif prices[k] < prices[i]:
                i = j = k
        
        return res
            

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 1) {
            return 1;
        }
        int res = 0;
        int l = 0, r = 0;

        Map<Character, Integer> hash = new HashMap<>();
        while (r < s.length()) {
            char curr = s.charAt(r);
            Integer pos = hash.get(curr);
            hash.put(curr, r);

            if (pos != null) {
                res = Math.max(res, r - l);
                if (pos >= l) {
                    l = pos + 1;
                }
            }
            r++;
        }

        return Math.max(res, r - l);
    }
}
